### PR TITLE
fix: refactor packaging

### PR DESCRIPTION
**Pull request by peternhale**
@W-11786892@
@W-11784677@
@W-11798867@

This PR must be merged in conjunction with a peer PR (not filed yet) on plugin-packaging

### CODE REVIEW

File: src/package/package.ts [hunk 1]
```diff
@@ -41,131 +46,206 @@ const messages = Messages.loadMessages('@salesforce/packaging', 'package');
...
+   * @param project - instance of SfProject
+   * @param options - see {@link PackageVersionListOptions}
+   */
+  public static async listVersions(
+    connection: Connection,
+    project: SfProject,
+    options?: PackageVersionListOptions
+  ): Promise<PackageVersionListResult[]> {
+    // resolve/verify packages
+    const packages = options?.packages.map((pkg) => {
```
Comment (WillieRuemmele) [inline src/package/package.ts:108]: `packages?.map` as `packages` could be undefined too

File: src/package/packageVersionList.ts [hunk 2]
```diff
@@ -35,6 +35,7 @@ const logger = Logger.childFromRoot('packageVersionList');
 export async function listPackageVersions(
   options: ListPackageVersionOptions
 ): Promise<QueryResult<PackageVersionListResult>> {
+  // TODO: replaces with connection.autoFetchQuery w/ tooling enabled
```
Comment (WillieRuemmele) [inline src/package/packageVersionList.ts:38]: has this been added to core?

File: src/interfaces/packagingInterfacesAndType.ts [hunk 3]
```diff
@@ -435,3 +418,22 @@ export const Package1VersionEvents = {
...
+  'SubscriberPackageId',
+  'Name',
+  'Description',
+  'NamespacePrefix',
+  'ContainerOptions',
+  'IsDeprecated',
+  'IsOrgDependent',
+  'ConvertedFromPackageId',
+  'PackageErrorUsername',
+];
```
Comment (shetzel) [inline src/interfaces/packagingInterfacesAndType.ts:439]: nit: this isn't a type

File: src/package/package.ts [hunk 4]
```diff
@@ -41,131 +46,206 @@ const messages = Messages.loadMessages('@salesforce/packaging', 'package');
...
  * This class provides the base implementation for a package.
+ * To create a new instance of a package, use the static async Package.create({connection, project, packageOrAliasId}) method.
  */
-export class Package extends AsyncCreatable<PackageOptions> implements IPackage {
+export class Package {
+  private packageId: string;
+  private packageData: PackagingSObjects.Package2;
   public constructor(private options: PackageOptions) {
-    super(options);
+    this.init();
```
Comment (shetzel) [inline src/package/package.ts:55]: could just move the code from `init` to the constructor.

File: src/package/package.ts [hunk 5]
```diff
@@ -41,131 +46,206 @@ const messages = Messages.loadMessages('@salesforce/packaging', 'package');
...
-  public static validateId(id: string, type: PackageIdType): void {
-    const prefix = packagePrefixes[type];
-    if (!id.startsWith(prefix)) {
-      throw messages.createError('invalidPackageId', [type, id, prefix]);
-    }
-    if (!sfdc.validateSalesforceId(id)) {
-      throw messages.createError('invalidIdLength', [type, id]);
-    }
+  public static async create(options: PackageOptions): Promise<Package> {
+    return new Package(options);
```
Comment (shetzel) [inline src/package/package.ts:64]: If we want to be consistent with `PackageVersion` then `createPackage` code should move here and that function goes away.

File: src/package/package.ts [hunk 6]
```diff
@@ -41,131 +46,206 @@ const messages = Messages.loadMessages('@salesforce/packaging', 'package');
...
+   * @param options
+   * @param project
+   */
+  public static async convert(
+    pkgId: string,
+    connection: Connection,
+    options: ConvertPackageOptions,
+    project?: SfProject
+  ): Promise<PackageVersionCreateRequestResult> {
+    return await convertPackage(pkgId, connection, options, project);
```
Comment (shetzel) [inline src/package/package.ts:140]:
```suggestion
    return convertPackage(pkgId, connection, options, project);
```

**Review by shetzel — COMMENTED** [review]
(partial review)

File: src/package/package.ts [hunk 7]
```diff
@@ -41,131 +46,206 @@ const messages = Messages.loadMessages('@salesforce/packaging', 'package');
...
-    return await uninstallPackage(id, this.options.connection, wait);
+  /**
+   * Returns the list of package versions for the package.
+   * See {@link PackageVersionListOptions} for list options
+   *
+   * @param options
+   * @returns {Promise<PackageVersionListResult[]>}
+   */
+  public async getPackageVersions(options?: PackageVersionListOptions): Promise<PackageVersionListResult[]> {
+    // This should be calling PackageVersion.list() here, but that method is not implemented yet.
```
[the remark this answers is not in the document]
Comment (peternhale) [reply src/package/package.ts:169]: @WillieRuemmele yes. I moved getPackageVersions from PV class to package class.

File: src/interfaces/packagingInterfacesAndType.ts (hunk 3 above)
Comment (peternhale) [reply src/interfaces/packagingInterfacesAndType.ts:439]: OK, will move to package class

File: src/package/package.ts [hunk 8]
```diff
@@ -37,135 +41,216 @@ const packagePrefixes = {
...
-      });
-      const errHeader = errors.length > 0 ? `\n=== Errors\n${errors.join('\n')}` : '';
-      const err = messages.getMessage('defaultErrorMessage', [id, result.Id]);
-
-      throw new SfError(`${err}${errHeader}`, 'UNINSTALL_ERROR', [messages.getMessage('action')]);
+  public async delete(): Promise<PackageSaveResult> {
+    const updateResult = await this.options.connection.tooling.delete('Package2', [this.packageId]).catch((err) => {
+      throw applyErrorAction(massageErrorMessage(err as Error));
+    });
+    if (updateResult[0]?.success) {
```
Comment (WillieRuemmele) [inline src/package/package.ts:210]: should this be `!updateResult[0]?.success`? If it's correct maybe adding a comment why a successful delete throws an error

Comment (peternhale) [reply src/package/package.ts:210]: The catch of line 207 handles any fatal errors that occur while processing the delete. The condition checks if the delete operation was successful. This code will be changing on next push.

File: test/package/packageVersionCreate.test.ts [hunk 9]
```diff
@@ -10,7 +10,7 @@ import { instantiateContext, MockTestOrgData, restoreContext, stubContext } from
 import { expect } from 'chai';
 import { Connection, SfProject } from '@salesforce/core';
 import * as xml2js from 'xml2js';
-import * as pkgUtils from '../../src/utils/packageUtils';
+// import * as pkgUtils from '../../src/utils/packageUtils';
```
Comment (shetzel) [inline test/package/packageVersionCreate.test.ts:13]: Can we remove the commented code on lines 13, 25, and 64 now?

**Review by shetzel — APPROVED** [review]

**Review by WillieRuemmele — APPROVED** [review]

**Pull request closed without merging**